File: source/modulo_new_core/src/translators/WriteStateConversion.cpp
#include "modulo_new_core/translators/WriteStateConversion.hpp"

#include <state_representation/exceptions/EmptyStateException.hpp>
#include <state_representation/space/cartesian/CartesianPose.hpp>

using namespace state_representation;

namespace modulo_new_core::translators {

static void write_point(geometry_msgs::msg::Point& msg, const Eigen::Vector3d& vector) {
  msg.x = vector.x();
  msg.y = vector.y();
  msg.z = vector.z();
}

static void write_vector3(geometry_msgs::msg::Vector3& msg, const Eigen::Vector3d& vector) {
  msg.x = vector.x();
  msg.y = vector.y();
  msg.z = vector.z();
}

static void write_quaternion(geometry_msgs::msg::Quaternion& msg, const Eigen::Quaterniond& quat) {
  msg.w = quat.w();
  msg.x = quat.x();
  msg.y = quat.y();
  msg.z = quat.z();
}

void write_msg(geometry_msgs::msg::Accel& msg, const CartesianState& state, const rclcpp::Time&) {
  if (state.is_empty()) {
    throw exceptions::EmptyStateException(state.get_name() + " state is empty while attempting to publish it");
  }
  write_vector3(msg.linear, state.get_linear_acceleration());
  write_vector3(msg.angular, state.get_angular_acceleration());
}

void write_msg(geometry_msgs::msg::AccelStamped& msg, const CartesianState& state, const rclcpp::Time& time) {
  write_msg(msg.accel, state, time);
  msg.header.stamp = time;
  msg.header.frame_id = state.get_reference_frame();
}

void write_msg(geometry_msgs::msg::Pose& msg, const CartesianState& state, const rclcpp::Time&) {
  if (state.is_empty()) {
    throw exceptions::EmptyStateException(state.get_name() + " state is empty while attempting to publish it");
  }
  write_point(msg.position, state.get_position());
  write_quaternion(msg.orientation, state.get_orientation());
}

void write_msg(geometry_msgs::msg::PoseStamped& msg, const CartesianState& state, const rclcpp::Time& time) {
  write_msg(msg.pose, state, time);
  msg.header.stamp = time;
  msg.header.frame_id = state.get_reference_frame();
}

void write_msg(geometry_msgs::msg::Transform& msg, const CartesianState& state, const rclcpp::Time&) {
  if (state.is_empty()) {
    throw exceptions::EmptyStateException(state.get_name() + " state is empty while attempting to publish it");
  }
  write_vector3(msg.translation, state.get_position());
  write_quaternion(msg.rotation, state.get_orientation());
}

void write_msg(geometry_msgs::msg::TransformStamped& msg, const CartesianState& state, const rclcpp::Time& time) {
  write_msg(msg.transform, state, time);
  msg.header.stamp = time;
  msg.header.frame_id = state.get_reference_frame();
  msg.child_frame_id = state.get_name();
}

void write_msg(geometry_msgs::msg::Twist& msg, const CartesianState& state, const rclcpp::Time&) {
  if (state.is_empty()) {
    throw exceptions::EmptyStateException(state.get_name() + " state is empty while attempting to publish it");
  }
  write_vector3(msg.linear, state.get_linear_velocity());
  write_vector3(msg.angular, state.get_angular_velocity());
}

void write_msg(geometry_msgs::msg::TwistStamped& msg, const CartesianState& state, const rclcpp::Time& time) {
  write_msg(msg.twist, state, time);
  msg.header.stamp = time;
  msg.header.frame_id = state.get_reference_frame();
}

void write_msg(geometry_msgs::msg::Wrench& msg, const CartesianState& state, const rclcpp::Time&) {
  if (state.is_empty()) {
    throw exceptions::EmptyStateException(state.get_name() + " state is empty while attempting to publish it");
  }
  write_vector3(msg.force, state.get_force());
  write_vector3(msg.torque, state.get_torque());
}

void write_msg(geometry_msgs::msg::WrenchStamped& msg, const CartesianState& state, const rclcpp::Time& time) {
  write_msg(msg.wrench, state, time);
  msg.header.stamp = time;
  msg.header.frame_id = state.get_reference_frame();
}

void write_msg(sensor_msgs::msg::JointState& msg, const JointState& state, const rclcpp::Time& time) {
  if (state.is_empty()) {
    throw exceptions::EmptyStateException(state.get_name() + " state is empty while attempting to publish it");
  }
  msg.header.stamp = time;
  msg.name = state.get_names();
  msg.position = std::vector<double>(state.get_positions().data(), state.get_positions().data() + state.get_positions().size());
  msg.velocity = std::vector<double>(state.get_velocities().data(), state.get_velocities().data() + state.get_velocities().size());
  msg.effort = std::vector<double>(state.get_torques().data(), state.get_torques().data() + state.get_torques().size());
}

void write_msg(tf2_msgs::msg::TFMessage& msg, const CartesianState& state, const rclcpp::Time& time) {
  if (state.is_empty()) {
    throw exceptions::EmptyStateException(state.get_name() + " state is empty while attempting to publish it");
  }
  geometry_msgs::msg::TransformStamped transform;
  write_msg(transform, state, time);
  msg.transforms.push_back(transform);
}

template <typename U, typename T>
void write_msg(U& msg, const Parameter<T>& state, const rclcpp::Time&) {
  if (state.is_empty()) {
    throw exceptions::EmptyStateException(state.get_name() + " state is empty while attempting to publish it");
  }
  msg.data = state.get_value();
}

template void write_msg<std_msgs::msg::Float64, double>(std_msgs::msg::Float64& msg, const Parameter<double>& state, const rclcpp::Time&);

template void write_msg<std_msgs::msg::Float64MultiArray, std::vector<double>>(std_msgs::msg::Float64MultiArray& msg, const Parameter<std::vector<double>>& state, const rclcpp::Time&);

template void write_msg<std_msgs::msg::Bool, bool>(std_msgs::msg::Bool& msg, const Parameter<bool>& state, const rclcpp::Time&);

template void write_msg<std_msgs::msg::String, std::string>(std_msgs::msg::String& msg, const Parameter<std::string>& state, const rclcpp::Time&);

template <>
void write_msg(geometry_msgs::msg::Transform& msg, const Parameter<CartesianPose>& state, const rclcpp::Time& time) {
  write_msg(msg, state.get_value(), time);
}

template <>
void write_msg(geometry_msgs::msg::TransformStamped& msg, const Parameter<CartesianPose>& state, const rclcpp::Time& time) {
  write_msg(msg, state.get_value(), time);
}

template <>
void write_msg(tf2_msgs::msg::TFMessage& msg, const Parameter<CartesianPose>& state, const rclcpp::Time& time) {
  write_msg(msg, state.get_value(), time);
}

void write_msg(std_msgs::msg::Bool& msg, const bool& state, const rclcpp::Time&) {
  msg.data = state;
}

void write_msg(std_msgs::msg::Float64& msg, const double& state, const rclcpp::Time&) {
  msg.data = state;
}

void write_msg(std_msgs::msg::Float64MultiArray& msg, const std::vector<double>& state, const rclcpp::Time&) {
  msg.data = state;
}

void write_msg(std_msgs::msg::Int32& msg, const int& state, const rclcpp::Time&) {
  msg.data = state;
}

void write_msg(std_msgs::msg::String& msg, const std::string& state, const rclcpp::Time&) {
  msg.data = state;
}
}// namespace modulo_new_core::translators

File: source/modulo_new_core/modulo_new_core/translators/write_state_conversion.py
import clproto
import geometry_msgs.msg as geometry
import numpy as np
import rclpy.time
import sensor_msgs.msg
import state_representation as sr
from modulo_new_core.encoded_state import EncodedState


def write_point(vector: np.array) -> geometry.Point:
    msg = geometry.Point()
    msg.x = vector[0]
    msg.y = vector[1]
    msg.z = vector[2]
    return msg


def write_vector(vector: np.array) -> geometry.Vector3:
    msg = geometry.Vector3()
    msg.x = vector[0]
    msg.y = vector[1]
    msg.z = vector[2]
    return msg


def write_quaternion(quat: np.array) -> geometry.Quaternion:
    msg = geometry.Quaternion()
    msg.w = quat[0]
    msg.x = quat[1]
    msg.y = quat[2]
    msg.z = quat[3]
    return msg


def write_msg(msg, state):
    if isinstance(msg, geometry.Accel) and isinstance(state, sr.CartesianState):
        msg.linear = write_vector(state.get_linear_acceleration())
        msg.angular = write_vector(state.get_angular_acceleration())
    elif isinstance(msg, geometry.Pose) and isinstance(state, sr.CartesianState):
        msg.position = write_point(state.get_position())
        msg.orientation = write_quaternion(state.get_orientation())
    elif isinstance(msg, geometry.Transform) and isinstance(state, sr.CartesianState):
        msg.translation = write_vector(state.get_position())
        msg.rotation = write_quaternion(state.get_orientation())
    elif isinstance(msg, geometry.Twist) and isinstance(state, sr.CartesianState):
        msg.linear = write_vector(state.get_linear_velocity())
        msg.angular = write_vector(state.get_angular_velocity())
    elif isinstance(msg, geometry.Wrench) and isinstance(state, sr.CartesianState):
        msg.force = write_vector(state.get_force())
        msg.torque = write_vector(state.get_torque())
    elif isinstance(msg, sensor_msgs.msg.JointState) and isinstance(state, sr.JointState):
        msg.name = state.get_names()
        msg.position = state.get_positions().tolist()
        msg.velocity = state.get_velocities().tolist()
        msg.effort = state.get_torques().tolist()
    else:
        raise RuntimeError("This message type is not supported")
    return msg


def write_msg_stamped(msg, state, time: rclpy.time.Time):
    if isinstance(msg, geometry.AccelStamped):
        write_msg(msg.accel, state)
        msg.header.stamp = time
        msg.header.frame_id = state.get_reference_frame()
    elif isinstance(msg, geometry.PoseStamped):
        write_msg(msg.pose, state)
        msg.header.stamp = time
        msg.header.frame_id = state.get_reference_frame()
    elif isinstance(msg, geometry.TransformStamped):
        write_msg(msg.transform, state)
        msg.header.stamp = time
        msg.header.frame_id = state.get_reference_frame()
        msg.child_frame_id = state.get_name()
    elif isinstance(msg, geometry.TwistStamped):
        write_msg(msg.twist, state)
        msg.header.stamp = time
        msg.header.frame_id = state.get_reference_frame()
    elif isinstance(msg, geometry.WrenchStamped):
        write_msg(msg.wrench, state)
        msg.header.stamp = time
        msg.header.frame_id = state.get_reference_frame()
    else:
        raise RuntimeError("This message type is not supported")
    return state


def write_clproto_msg(state: sr.State, msg_type: clproto.MessageType) -> EncodedState():
    msg = EncodedState()
    if isinstance(state, sr.State) and isinstance(msg_type, clproto.MessageType):
        msg.data = clproto.encode(state, msg_type)
    else:
        raise RuntimeError("Your inputs do not have the correct types")
    return msg
